src/HappyEyeBallsConnectionBuilder.php

```diff
@@ -0,0 +1,304 @@
+<?php
+
+namespace React\Socket;
+
+use React\Dns\Model\Message;
+use React\Dns\Resolver\Resolver;
+use React\EventLoop\LoopInterface;
+use React\EventLoop\TimerInterface;
+use React\Promise;
+use React\Promise\CancellablePromiseInterface;
+use InvalidArgumentException;
+use RuntimeException;
+
+/**
+ * @internal
+ */
+final class HappyEyeBallsConnectionBuilder
+{
+    public $loop;
+    public $connector;
+    public $resolver;
+    public $uri;
+    public $host;
+    public $resolved = array();
+    public $resolverPromises = array();
+    public $connectionPromises = array();
+    public $connectQueue = array();
+    public $timer;
+    public $parts;
+    public $ipsCount = 0;
+    public $failureCount = 0;
+    public $resolve;
+    public $reject;
+
+    public static function create(LoopInterface $loop, ConnectorInterface $connector, Resolver $resolver, $uri)
+    {
+        $connectionBuilder = new self($loop, $connector, $resolver, $uri);
+
+        return $connectionBuilder->connect();
+    }
+
+    private function __construct(LoopInterface $loop, ConnectorInterface $connector, Resolver $resolver, $uri)
+    {
+        $this->loop = $loop;
+        $this->connector = $connector;
+        $this->resolver = $resolver;
+        $this->uri = $uri;
+
+        if (\strpos($uri, '://') === false) {
+            $this->parts = \parse_url('tcp://' . $uri);
+            unset($this->parts['scheme']);
+        } else {
+            $this->parts = \parse_url($uri);
+        }
+
+        if (!$this->parts || !isset($this->parts['host'])) {
+            throw new \InvalidArgumentException('Given URI "' . $uri . '" is invalid');
+        }
+
+        $this->host = \trim($this->parts['host'], '[]');
+    }
+
+    /**
+     * @internal
+     */
+    public function connect()
+    {
+        // skip DNS lookup / URI manipulation if this URI already contains an IP
+        if (false !== \filter_var($this->host, \FILTER_VALIDATE_IP)) {
+            return $this->connector->connect($this->uri);
+        }
```

Comment:
Perhaps move this logic closer to the beginning of the execution flow to skip this entire class?

Reply:
Done :+1: